Automattic/woocommerce-payments
# Move Promise instantiation into memoized function to ensure caching consistency

Fixes https://github.com/Automattic/woocommerce-payments/issues/9851

#### Changes proposed in this Pull Request
It was discovered that for some reason, `canMakePayment` for payment methods (non-express checkout payment methods) is executed only upon page load, but not executed on subsequent actions like e.g. billing country update. This applies to Blocks checkout only. 

The side effect ended up to be in the `canMakePayment`'s [callback implementation for express checkout methods](https://github.com/Automattic/woocommerce-payments/blob/develop/client/express-checkout/utils/checkPaymentMethodIsAvailable.js#L16). There are two aspects worth mentioning:

1️⃣  WooCommerce Blocks [takes all the registered payment methods](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/data/payment/utils/check-payment-methods.ts#L157-L159) and then [calls `canMakePayment`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/data/payment/utils/check-payment-methods.ts#L223) on each of those. Typically, express checkout payment methods are evaluated first due to the default registration order, unless explicitly configured otherwise.

2️⃣  The `canMakePayment`'s [callback implementation for express checkout methods](https://github.com/Automattic/woocommerce-payments/blob/develop/client/express-checkout/utils/checkPaymentMethodIsAvailable.js#L16) uses `memoize`, which caches the resolved result (a boolean) from the first execution. Subsequent calls to this function return the cached boolean instead of a promise with some resolution state, leaving promises unresolved in those cases. This behavior caused [the promises](https://github.com/Automattic/woocommerce-payments/blob/develop/client/express-checkout/blocks/index.js#L42-L44) to remain unhandled, as the cache bypassed the promise resolution logic.

This PR resolves the issue by caching the promise itself instead of its resolved value. Previously, memoize cached the resolved boolean, causing subsequent calls to bypass the promise logic. Now, the memoized function returns and caches the promise, ensuring that subsequent calls return the cached resolved promise which is handled by Blocks.













#### Testing instructions





For Blocks checkout

* check on `develop` that dynamical filtering upon billing country change is not working
* check out `fix/unhandled-promises` and make sure that dynamical filtering of payment methods works 
    * enable e.g. Klarna
    * while opening the page, confirm this payment method is visible or invisible (depending on the country selected) 
    * switch the country and confirm Klarna is appearing and disappearing as expected depending on the country
* confirm that the testing requirements from https://github.com/Automattic/woocommerce-payments/pull/9378 are still met

-------------------


**Post merge**
